### PR TITLE
New metric: CLIP IQA

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

Adds `CLIPImageQualityAssessment` as new metric to the `image` subpackage.

<details>
  <summary>Before submitting</summary>

- [ ] Was this **discussed/agreed** via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/torchmetrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to **update the docs**?
- [ ] Did you write any new **necessary tests**?

</details>

<details>
  <summary>PR review</summary>

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

</details>

## Did you have fun?

Make sure you had fun coding 🙃

----
:books: Documentation preview :books:: https://torchmetrics--1931.org.readthedocs.build/en/1931/

### CODE REVIEW

**Review by Borda — APPROVED** [review]

**Review by lantiga — APPROVED** [review]
Looks great!

I'm a bit unclear about dependencies. CLIP here comes from transformers, but there's no such dependency in `requirements/image.txt` (while it's there for `text.txt` and `multimodal.txt`).

To clarify, I wouldn't advise to add transformers to `image.txt` just for this one metric, but I'm wondering if we should move this metric under `multimodal`, since the clip metric is also there.

**Comment by SkafteNicki** [comment]
@lantiga thanks for the review. That is a really good point, so let me just move in to multimodal instead of image.

File: src/torchmetrics/functional/multimodal/clip_iqa.py [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+        "openai/clip-vit-large-patch14",
+    ]
+) -> Tuple[_CLIPModel, _CLIPProcessor]:
+    """Extract the CLIP model and processor from the model name or path."""
+    if model_name_or_path == "clip_iqa":
+        if not _PIQ_GREATER_EQUAL_0_8:
+            raise ValueError(
+                "For metric `clip_iqa` to work with argument `model_name_or_path` set to default value `'clip_iqa'`"
+                ", package `piq` version v0.8.0 or later must be installed. Either install with `pip install piq` or"
+                "`pip install torchmetrics[image]`"
```
Comment (stancld) [inline src/torchmetrics/functional/multimodal/clip_iqa.py:77]:
```suggestion
                "`pip install torchmetrics[multimodal]`"
```

File: src/torchmetrics/functional/multimodal/clip_iqa.py [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+
+        model = piq.clip_iqa.clip.load().eval()
+        # any model checkpoint can be used here because the tokenizer is the same for all
+        processor = _CLIPProcessor.from_pretrained("openai/clip-vit-base-patch16")
+        return model, processor
+    return _get_clip_model_and_processor(model_name_or_path)
+
+
+def _clip_iqa_format_prompts(prompts: Tuple[Union[str, Tuple[str, str]]] = ("quality",)) -> Tuple[List[str], List[str]]:
+    """Converts the provided keywords into a list of prompts for the model to calculate the achor vectors.
```
Comment (stancld) [inline src/torchmetrics/functional/multimodal/clip_iqa.py:90]:
```suggestion
    """Converts the provided keywords into a list of prompts for the model to calculate the anchor vectors.
```

File: src/torchmetrics/functional/multimodal/clip_iqa.py [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+
+    Args:
+        prompts: A string, list of strings or tuple of strings. If a string is provided, it must be one of the
+            availble prompts. If a list of strings is provided, all strings must be one of the availble prompts.
+            If a tuple of strings is provided, it must be of length 2 and the first string must be a positive prompt
+            and the second string must be a negative prompt.
```
Comment (stancld) [inline src/torchmetrics/functional/multimodal/clip_iqa.py:96]: This looks a bit inconsistent with type annotation which expects `Tuple[str | Tuple[str, str]]`

File: src/torchmetrics/multimodal/clip_iqa.py [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+
+    Raises:
+        ModuleNotFoundError:
+            If transformers package is not installed or version is lower than 4.10.0
+        ValueError:
+            If promts is a tuple and it is not of length 2
+        ValueError:
+            If promts is a string and it is not one of the available promts
+        ValueError:
+            If promts is a list of strings and not all strings are one of the available promts
```
Comment (stancld) [inline src/torchmetrics/multimodal/clip_iqa.py:125]:
```suggestion
            If `prompts` is a tuple and it is not of length 2
        ValueError:
            If `prompts` is a string and it is not one of the available prompts
        ValueError:
            If `prompts` is a list of strings and not all strings are one of the available prompts
```

File: src/torchmetrics/multimodal/clip_iqa.py [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+
+    Example::
+        Single promt:
```
Comment (stancld) [inline src/torchmetrics/multimodal/clip_iqa.py:128]:
```suggestion
        Single prompt:
```

File: src/torchmetrics/multimodal/clip_iqa.py [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+        >>> from torchmetrics.image import CLIPImageQualityAssessment
+        >>> import torch
+        >>> _ = torch.manual_seed(42)
+        >>> imgs = torch.randint(255, (2, 3, 224, 224)).float()
+        >>> metric = CLIPImageQualityAssessment()
+        >>> metric(imgs)
+        tensor([0.8894, 0.8902])
+
+    Example::
+        Multiple promts:
```
Comment (stancld) [inline src/torchmetrics/multimodal/clip_iqa.py:139]:
```suggestion
        Multiple prompts:
```

File: src/torchmetrics/multimodal/clip_iqa.py [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+        >>> from torchmetrics.image import CLIPImageQualityAssessment
+        >>> import torch
+        >>> _ = torch.manual_seed(42)
+        >>> imgs = torch.randint(255, (2, 3, 224, 224)).float()
+        >>> metric = CLIPImageQualityAssessment(prompts=("quality", "brightness"))
+        >>> metric(imgs)
+        {'quality': tensor([0.8894, 0.8902]), 'brightness': tensor([0.5507, 0.5208])}
+
+    Example::
+        Custom promts. Must always be a tuple of length 2, with a positive and negative prompt.
```
Comment (stancld) [inline src/torchmetrics/multimodal/clip_iqa.py:150]:
```suggestion
        Custom prompts. Must always be a tuple of length 2, with a positive and negative prompt.
```

File: src/torchmetrics/functional/multimodal/clip_iqa.py [hunk 8]
```diff
@@ -0,0 +1,326 @@
...
+    model: _CLIPModel,
+    processor: _CLIPProcessor,
+    data_range: Union[int, float],
+    device: Union[str, torch.device],
+) -> Tensor:
+    images = images / float(data_range)
+    """Update function for CLIP IQA."""
+    if model_name_or_path == "clip_iqa":
+        default_mean = torch.tensor([0.48145466, 0.4578275, 0.40821073], device=device).view(1, 3, 1, 1)
+        default_std = torch.tensor([0.26862954, 0.26130258, 0.27577711], device=device).view(1, 3, 1, 1)
```
Comment (stancld) [inline src/torchmetrics/functional/multimodal/clip_iqa.py:187]: Those values are chosen based on some original implementation? 🤔 Maybe shouldn't we add some reference to the paper on how they were derived?

File: src/torchmetrics/functional/multimodal/clip_iqa.py [hunk 9]
```diff
@@ -0,0 +1,326 @@
...
+    prompts: Tuple[Union[str, Tuple[str, str]]] = ("quality",),
+) -> Union[Tensor, Dict[str, Tensor]]:
+    """Calculates `CLIP-IQA`_, that can be used to measure the visual content of images.
+
+    The metric is based on the `CLIP`_ model, which is a neural network trained on a variety of (image, text) pairs to
+    be able to generate a vector representation of the image and the text that is similar if the image and text are
+    semantically similar.
+
+    The metric works by calculating the cosine similarity between user provided images and pre-defined promts. The
+    promts always comes in pairs of "positive" and "negative" such as "Good photo." and "Bad photo.". By calculating
```
Comment (stancld) [inline src/torchmetrics/functional/multimodal/clip_iqa.py:231]:
```suggestion
    prompts always come in pairs of "positive" and "negative" such as "Good photo." and "Bad photo.". By calculating
```

File: src/torchmetrics/functional/multimodal/clip_iqa.py [hunk 10]
```diff
@@ -0,0 +1,326 @@
...
+        * scary: "Scary photo." vs "Peaceful photo."
+        * new: "New photo." vs "Old photo."
+        * warm: "Warm photo." vs "Cold photo."
+        * real: "Real photo." vs "Abstract photo."
+        * beutiful: "Beautiful photo." vs "Ugly photo."
+        * lonely: "Lonely photo." vs "Sociable photo."
+        * relaxing: "Relaxing photo." vs "Stressful photo."
+
+    Args:
+        images: Either a single [N, C, H, W] tensor or a list of [C, H, W] tensors
```
Comment (stancld) [inline src/torchmetrics/functional/multimodal/clip_iqa.py:255]:
```suggestion
        images: Either a single ``[N, C, H, W]`` tensor or a list of ``[C, H, W]`` tensors
```

File: src/torchmetrics/functional/multimodal/clip_iqa.py [hunk 11]
```diff
@@ -0,0 +1,326 @@
...
+            and `"openai/clip-vit-large-patch14"`
+        data_range: The maximum value of the input tensor. For example, if the input images are in range [0, 255],
+            data_range should be 255. The images are normalized by this value.
+        prompts: A string, tuple of strings or nested tuple of strings. If a string is provided, it must be one of the
+            availble prompts. If a tuple of strings is provided, all strings must be one of the availble prompts.
+            If a nested tuple of strings is provided, it must be of length 2 and the first string must be a positive
+            prompt and the second string must be a negative prompt.
+
+    .. note:: If using the default `clip_iqa` model, the package `piq` must be installed. Either install with
+        `pip install piq` or `pip install torchmetrics[image]`.
```
Comment (stancld) [inline src/torchmetrics/functional/multimodal/clip_iqa.py:268]:
```suggestion
        `pip install piq` or `pip install torchmetrics[multimodal]`.
```

**Review by stancld — COMMENTED** [review]
Just left a few comments. Looks good! Thanks for this addition! 🚀
